### PR TITLE
HPCC-22296 SoapCall has extra / in first line of request

### PR DESCRIPTION
**Pull request by mayx**
- Clean the path part of the url by removing extra slashes when parsing
  the url.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mayx** [comment]
@jakesmith please review...

File: common/thorhelper/thorsoapcall.cpp [hunk 1]
```diff
@@ -170,6 +170,23 @@ class Url : public CInterface, implements IInterface
         return;
     }
 
+    void cleanPath()
+    {
+        int len = path.length();
```
Comment (jakesmith) [inline common/thorhelper/thorsoapcall.cpp]: picky: should be an unsigned (size32_t)

File: common/thorhelper/thorsoapcall.cpp [hunk 2]
```diff
@@ -170,6 +170,23 @@ class Url : public CInterface, implements IInterface
...
+        if (len <= 1)
+            return;
+        int gap = 0;
+        for (int i = 1; i < len; i++)
```
Comment (jakesmith) [inline common/thorhelper/thorsoapcall.cpp]: picky: int's should be unsigned

File: common/thorhelper/thorsoapcall.cpp [hunk 3]
```diff
@@ -170,6 +170,23 @@ class Url : public CInterface, implements IInterface
         return;
     }
 
+    void cleanPath()
```
Comment (jakesmith) [inline common/thorhelper/thorsoapcall.cpp]: It looks like cleanPath() is trying to walk the whole string and remove any duplicated '/' 's?
e.g. convert : "POST //WsPackageProcess/myxpath//sub HTTP/1.1"
to "POST /WsPackageProcess/myxpath/sub HTTP/1.1

Is that ok/right?

Could do with brief comment here.

File: common/thorhelper/thorsoapcall.cpp [hunk 4]
```diff
@@ -170,6 +170,23 @@ class Url : public CInterface, implements IInterface
...
+        int len = path.length();
+        if (len <= 1)
+            return;
+        int gap = 0;
+        for (int i = 1; i < len; i++)
+        {
+            if (path.charAt(i) == '/' && path.charAt(i-1-gap) == '/')
+                gap++;
+            else if (gap > 0)
+                path.setCharAt(i-gap, path.charAt(i));
```
Comment (jakesmith) [inline common/thorhelper/thorsoapcall.cpp]: may not be critical, but calling setCharAt per character is not particularly efficient.
If expected to be hit a lot, would be better to deal with underlying string and reset 'path' to new string when done.
e.g. could detach() and work on const char *writeptr

**Comment by jakesmith** [comment]
@mayx - a couple of comments/questions.

File: common/thorhelper/thorsoapcall.cpp (hunk 3 above)
Comment (mayx) [reply common/thorhelper/thorsoapcall.cpp]: Yes all duplicate slashes should be removed, at the beginning, in the middle or at the end of the path...

**Comment by mayx** [comment]
@jakesmith changes made per suggestion, please review again...

File: common/thorhelper/thorsoapcall.cpp [hunk 5]
```diff
@@ -170,6 +170,24 @@ class Url : public CInterface, implements IInterface
...
+    {
+        size32_t len = path.length();
+        if (len <= 1)
+            return;
+        size32_t gap = 0;
+        char* buf = path.detach();
+        //Remove all duplicate slashes
+        for (size32_t i = 1; i < len; i++)
+        {
+            if (buf[i] == '/' && buf[i-1-gap] == '/')
```
Comment (jakesmith) [inline common/thorhelper/thorsoapcall.cpp]: minor: [i-1-gap] is always the 1st '/' in a sequence, so you could code this more efficiently as a while loop.

Personally I'd code this routine something like this:
```
void cleanPath(const char *src, StringBuffer &output)
{
    const char *srcPtr = src;
    char *dst = output.reserveTruncate(strlen(src));
    char *dstPtr = dst;
    while (*srcPtr)
    {
        *dstPtr++ = *srcPtr;
        if ('/' == *srcPtr++)
        {
            while ('/' == *srcPtr)
                ++srcPtr;
        }
    }
    *dstPtr = '\0';
    output.setLength(dstPtr-dst);
}

```

You could pass a single StringBuffer as both the input and output, and detach() instead, but given the input in the Uri ctor isn't originally a StringBuffer (i.e. 'p' was just appended before calling cleanPath)...

**Comment by jakesmith** [comment]
@mayx - 1 comment/suggestion, but okay to be merged as is.

**Comment by mayx** [comment]
@jakesmith changes made per suggestion, please review again...

**Comment by jakesmith** [comment]
@mayx - looks fine.

@richardkchapman - ready to merge.
